## 1. Load Packages and Datasets

In [51]:
# Import packages
import pandas as pd
import re
import spacy
from collections import Counter
import nltk
from nltk import word_tokenize
from nltk.corpus import stopwords
from string import punctuation

In [52]:
train_df = pd.read_json('data/raw_data/train.json')
train_df.head()

document                                          full_text  \
0         7  Design Thinking for innovation reflexion-Avril...   
1        10  Diego Estrada\n\nDesign Thinking Assignment\n\...   
2        16  Reporting process\n\nby Gilberto Gamboa\n\nCha...   
3        20  Design Thinking for Innovation\n\nSindy Samaca...   
4        56  Assignment:  Visualization Reflection  Submitt...   

                                              tokens  \
0  [Design, Thinking, for, innovation, reflexion,...   
1  [Diego, Estrada, \n\n, Design, Thinking, Assig...   
2  [Reporting, process, \n\n, by, Gilberto, Gambo...   
3  [Design, Thinking, for, Innovation, \n\n, Sind...   
4  [Assignment, :,   , Visualization,  , Reflecti...   

                                 trailing_whitespace  \
0  [True, True, True, True, False, False, True, F...   
1  [True, False, False, True, True, False, False,...   
2  [True, False, False, True, True, False, False,...   
3  [True, True, True, False, False, True, False, ...   
4  [False, False, False, False, False, False, Fal...   

                                              labels  
0  [O, O, O, O, O, O, O, O, O, B-NAME_STUDENT, I-...  
1  [B-NAME_STUDENT, I-NAME_STUDENT, O, O, O, O, O...  
2  [O, O, O, O, B-NAME_STUDENT, I-NAME_STUDENT, O...  
3  [O, O, O, O, O, B-NAME_STUDENT, I-NAME_STUDENT...  
4  [O, O, O, O, O, O, O, O, O, O, O, O, B-NAME_ST...

In [53]:
test_df = pd.read_json('data/raw_data/test.json')
test_df.head()

document                                          full_text  \
0         7  Design Thinking for innovation reflexion-Avril...   
1        10  Diego Estrada\n\nDesign Thinking Assignment\n\...   
2        16  Reporting process\n\nby Gilberto Gamboa\n\nCha...   
3        20  Design Thinking for Innovation\n\nSindy Samaca...   
4        56  Assignment:  Visualization Reflection  Submitt...   

                                              tokens  \
0  [Design, Thinking, for, innovation, reflexion,...   
1  [Diego, Estrada, \n\n, Design, Thinking, Assig...   
2  [Reporting, process, \n\n, by, Gilberto, Gambo...   
3  [Design, Thinking, for, Innovation, \n\n, Sind...   
4  [Assignment, :,   , Visualization,  , Reflecti...   

                                 trailing_whitespace  
0  [True, True, True, True, False, False, True, F...  
1  [True, False, False, True, True, False, False,...  
2  [True, False, False, True, True, False, False,...  
3  [True, True, True, False, False, True, False, ...  
4  [False, False, False, False, False, False, Fal...

## 2. Exploratory Data Analysis

In [54]:
train_df.isna().sum()

document               0
full_text              0
tokens                 0
trailing_whitespace    0
labels                 0
dtype: int64

In [55]:
test_df.isna().sum()

document               0
full_text              0
tokens                 0
trailing_whitespace    0
dtype: int64

We are able to verify that neither dataset has null values that would need to be pre-processed.

In [56]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6807 entries, 0 to 6806
Data columns (total 5 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   document             6807 non-null   int64 
 1   full_text            6807 non-null   object
 2   tokens               6807 non-null   object
 3   trailing_whitespace  6807 non-null   object
 4   labels               6807 non-null   object
dtypes: int64(1), object(4)
memory usage: 266.0+ KB


In [57]:
test_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 4 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   document             10 non-null     int64 
 1   full_text            10 non-null     object
 2   tokens               10 non-null     object
 3   trailing_whitespace  10 non-null     object
dtypes: int64(1), object(3)
memory usage: 448.0+ bytes


We know that the train set has 6,807 rows of data and the test set only has 10 rows.

In [74]:
import pandas as pd
from string import punctuation
from nltk.corpus import stopwords

PUNCTUATION = set(punctuation)
STOPWORDS = set(stopwords.words("english"))

def lowercase(tokens):
    return [token.lower() for token in tokens]

def remove_stopwords(tokens):
    stop_words = set(stopwords.words("english"))
    return [token for token in tokens if token.lower() not in stop_words]

def remove_punctuation(tokens):
    return [''.join(ch for ch in token if ch not in PUNCTUATION) for token in tokens]

def remove_empty_and_newline(tokens):
    return [token for token in tokens if token.strip() != '' and token.strip() != ' \n\n']

def preprocess(tokens):
    tokens = lowercase(tokens)
    tokens = remove_punctuation(tokens)
    tokens = remove_stopwords(tokens)
    tokens = remove_empty_and_newline(tokens)
    return tokens

In [75]:
train_df['tokens_processed'] = train_df['tokens'].apply(preprocess)
train_df['tokens_count'] = train_df['tokens'].map(len)
train_df.head()

document                                          full_text  \
0         7  Design Thinking for innovation reflexion-Avril...   
1        10  Diego Estrada\n\nDesign Thinking Assignment\n\...   
2        16  Reporting process\n\nby Gilberto Gamboa\n\nCha...   
3        20  Design Thinking for Innovation\n\nSindy Samaca...   
4        56  Assignment:  Visualization Reflection  Submitt...   

                                              tokens  \
0  [Design, Thinking, for, innovation, reflexion,...   
1  [Diego, Estrada, \n\n, Design, Thinking, Assig...   
2  [Reporting, process, \n\n, by, Gilberto, Gambo...   
3  [Design, Thinking, for, Innovation, \n\n, Sind...   
4  [Assignment, :,   , Visualization,  , Reflecti...   

                                 trailing_whitespace  \
0  [True, True, True, True, False, False, True, F...   
1  [True, False, False, True, True, False, False,...   
2  [True, False, False, True, True, False, False,...   
3  [True, True, True, False, False, True, False, ...   
4  [False, False, False, False, False, False, Fal...   

                                              labels  \
0  [O, O, O, O, O, O, O, O, O, B-NAME_STUDENT, I-...   
1  [B-NAME_STUDENT, I-NAME_STUDENT, O, O, O, O, O...   
2  [O, O, O, O, B-NAME_STUDENT, I-NAME_STUDENT, O...   
3  [O, O, O, O, O, B-NAME_STUDENT, I-NAME_STUDENT...   
4  [O, O, O, O, O, O, O, O, O, O, O, O, B-NAME_ST...   

                                    tokens_processed  tokens_count  
0  [design, thinking, innovation, reflexion, avri...           753  
1  [diego, estrada, design, thinking, assignment,...           563  
2  [reporting, process, gilberto, gamboa, challen...           729  
3  [design, thinking, innovation, sindy, samaca, ...          1071  
4  [assignment, visualization, reflection, submit...          1927

In [91]:
test_df['tokens_processed'] = test_df['tokens'].apply(preprocess)
test_df['tokens_count'] = test_df['tokens'].map(len)
test_df.head()

document                                          full_text  \
0         7  Design Thinking for innovation reflexion-Avril...   
1        10  Diego Estrada\n\nDesign Thinking Assignment\n\...   
2        16  Reporting process\n\nby Gilberto Gamboa\n\nCha...   
3        20  Design Thinking for Innovation\n\nSindy Samaca...   
4        56  Assignment:  Visualization Reflection  Submitt...   

                                              tokens  \
0  [Design, Thinking, for, innovation, reflexion,...   
1  [Diego, Estrada, \n\n, Design, Thinking, Assig...   
2  [Reporting, process, \n\n, by, Gilberto, Gambo...   
3  [Design, Thinking, for, Innovation, \n\n, Sind...   
4  [Assignment, :,   , Visualization,  , Reflecti...   

                                 trailing_whitespace  \
0  [True, True, True, True, False, False, True, F...   
1  [True, False, False, True, True, False, False,...   
2  [True, False, False, True, True, False, False,...   
3  [True, True, True, False, False, True, False, ...   
4  [False, False, False, False, False, False, Fal...   

                                    tokens_processed  tokens_count  
0  [design, thinking, innovation, reflexion, avri...           753  
1  [diego, estrada, design, thinking, assignment,...           563  
2  [reporting, process, gilberto, gamboa, challen...           729  
3  [design, thinking, innovation, sindy, samaca, ...          1071  
4  [assignment, visualization, reflection, submit...          1927

In [86]:
def descriptive_stats(tokens, top_tokens=5, verbose=True):
    """
    Calculate descriptive statistics for a list of tokens.
    Args:
        tokens (list): List of tokens.
        top_tokens (int): Number of top tokens to display.
        verbose (bool): Print statistics if True.
    Returns:
        list: A list containing [num_tokens, num_unique_tokens, lexical_diversity, num_characters]
    """
    num_tokens = len(tokens)
    num_unique_tokens = len(set(tokens))
    lexical_diversity = num_unique_tokens / num_tokens
    num_characters = sum(len(token) for token in tokens)
    token_counter = Counter(tokens)
    most_common_tokens = token_counter.most_common(top_tokens)

    if verbose:
        print(f"Number of tokens: {num_tokens}")
        print(f"Number of unique tokens: {num_unique_tokens}")
        print(f"Number of characters: {num_characters}")
        print(f"Lexical diversity: {lexical_diversity:.3f}")
        print(f"Most common tokens (Top {top_tokens}):")
        for token, count in most_common_tokens:
            print(f" - {token}: {count}")

    return [num_tokens, num_unique_tokens, lexical_diversity, num_characters]

In [87]:
train_tokens = train_df['tokens_processed'].sum()
train_stats = descriptive_stats(train_tokens)

Number of tokens: 2225629
Number of unique tokens: 41804
Number of characters: 14598461
Lexical diversity: 0.019
Most common tokens (Top 5):
 - tool: 19479
 - team: 13522
 - design: 13260
 - would: 13077
 - challenge: 12704


In [92]:
test_tokens = test_df['tokens_processed'].sum()
test_stats = descriptive_stats(test_tokens)

Number of tokens: 3377
Number of unique tokens: 1501
Number of characters: 22651
Lexical diversity: 0.444
Most common tokens (Top 5):
 - tool: 34
 - would: 31
 - people: 29
 - money: 21
 - company: 20
